### PR TITLE
Fix: rewritting prey system

### PR DESCRIPTION
**Pull request by westwol**
The current prey system has flaws when it comes to stability, as there are people that know this bug and can crash this engine. I decided to rewrite the system on my free time, feel free to coolaborate.

Supported versions: 10.x and 11.49.

**99% completed.**

**New features**
- The system was safely written and will guarantee engine stability once all tests are completed.
- No database queries during sets/gets, all data is loaded upon log in and saved upon log out or server save.
- A list of monsters is generated depending on the player's level, avoiding to have noob monsters on high level players (fully configurable).
- Bonus reroll is as accurate as Tibia RL (steps).
- Monsters can't be repeated on different slots.

**Missing:**
- Client 11.8+ support.
- Remove old prey system functions.
- probably a few more things but not as important as these.

### CODE REVIEW

File: data/modules/scripts/prey system/prey.lua [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+local monsterList = {
+	[CONST_MONSTER_TIER_BRONZE] = {"Rat", "Rotworm", "Cave Rat", "Orc", "Dwarf", "Ghoul", "Minotaur", "Fire Elemental", "Earth Elemental"},
+	[CONST_MONSTER_TIER_SILVER] = {"Pirate Ghost", "Elder Wyrm", "Orclops Doomhauler", "Giant Spider", "Lizard Zaogun", "Nightmare"}, 
+	[CONST_MONSTER_TIER_GOLD] = {"Quara Pincher", "Quara Predator", "Ogre Brute", "Minotaur Amazon", "Guzzlemaw", "Grim Reaper"},
+	[CONST_MONSTER_TIER_PLATINUM] = {"Demon", "Plaguesmith", "Defiler", "Destroyer"}
+}
+
+function onRecvbyte(player, msg, byte)
+	if (byte == Prey.C_Packets.RequestData) then
```
Comment (gpedro) [inline data/modules/scripts/prey system/prey.lua]: this `(` and `)` is unnecessary.

File: data/modules/scripts/prey system/prey.lua [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+		player:sendPreyData(CONST_SLOT_FIRST)
+		player:sendPreyData(CONST_SLOT_SECOND)
+		player:sendPreyData(CONST_SLOT_THIRD)
+	elseif (byte == Prey.C_Packets.PreyAction) then
```
Comment (gpedro) [inline data/modules/scripts/prey system/prey.lua]: this `(` and `)` is unnecessary.

File: data/modules/scripts/prey system/prey.lua [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		print("TODO: Prey action")
+	end
+end
+
+function Player.getMonsterTier(self)
+	local tier = CONST_MONSTER_TIER_BRONZE
+	if (self:getLevel() > 0 and self:getLevel() < 50) then
+		tier = CONST_MONSTER_TIER_BRONZE
```
Comment (gpedro) [inline data/modules/scripts/prey system/prey.lua]: you don't need to allocate variable. just return the value.
```lua
if self:getLevel() > 0 and self:getLevel() < 50 then
  return CONST_MONSTER_TIER_BRONZE
end

if self:getLevel() >= 50 and self:getLevel() < 100 then
  return CONST_MONSTER_TIER_SILVER
end

if self:getLevel() >= 100 and self:getLevel() < 150 then
  return CONST_MONSTER_TIER_GOLD
end

if self:getLevel() >= 150 then
  return CONST_MONSTER_TIER_PLATINUM
end
```

**Review by mattyx14 — APPROVED** [review]

File: data/events/scripts/creature.lua [hunk 4]
```diff
@@ -212,5 +212,30 @@ function Creature:onDrainHealth(attacker, typePrimary, damagePrimary, typeSecond
...
+					break
+				end
+			end
+		end
+	-- New prey => Damage reduction
+	elseif (attacker:isMonster()) then
+		if (self:isPlayer()) then
+			for slot = CONST_PREY_SLOT_FIRST, CONST_PREY_SLOT_THIRD do
+				if (self:getPreyCurrentMonster(slot) == attacker:getName() and self:getPreyBonusType(slot) == CONST_BONUS_DAMAGE_REDUCTION) then
+					print("First damage: " .. damagePrimary)
```
Comment (gpedro) [inline data/events/scripts/creature.lua]: you forgot to remove debug messages xd

File: src/game.cpp [hunk 5]
```diff
@@ -4200,7 +4200,7 @@ bool Game::combatChangeHealth(Creature* attacker, Creature* target, CombatDamage
 		if (healthChange == 0) {
 			return true;
 		}
-
+
```
Comment (gpedro) [inline src/game.cpp:4203]: trailing spaces xd

File: src/iologindata.cpp [hunk 6]
```diff
@@ -222,10 +222,67 @@ bool IOLoginData::loadPlayerById(Player* player, uint32_t id)
...
+			debug << "MonsterList: " << player->preySlotMonsterList[slotNum] << " | ";
+			debug << "Free reroll in: " << player->preySlotFreeRerollIn[slotNum] << " | ";
+			debug << "Time-left: " << player->preySlotTimeLeft[slotNum] << " | ";
+			debug << "Next-use: " << player->preySlotNextUse[slotNum] << " | ";
+			debug << "Bonus type: " << player->preySlotBonusType[slotNum] << " | ";
+			debug << "Bonus value: " << player->preySlotBonusValue[slotNum] << " | ";
+			debug << "Bonus grade: " << player->preySlotBonusGrade[slotNum];
+
+			// Debuging
+			std::cout << debug.str() << std::endl;
```
Comment (gpedro) [inline src/iologindata.cpp]: you forgot to remove debug messages xd

**Review by mattyx14 — APPROVED** [review]

**Comment by westwol** [comment]
The system is done, we need people to test that it works as inteed.

Feel free to report bugs or suggestions.

**Comment by gpedro** [comment]
guys, can you test?
cc @FearLucien @pitis91 @Raphaellb @Olimpotibia

**Comment by Olimpotibia** [comment]
gpedro open new branch to test it if you can

ok is here: https://github.com/westwol/OTServBR-Global/tree/patch-3

**Pull request merged**